classes/village/VillageManager.php

```diff
@@ -1826,12 +1899,142 @@ public static function createBreakAllianceProposal(System $system, User $player,
+        if ($player->village->isEnemy($target_village_id)) {
+            return "Villages must be at peace in order to trade!";
+        }
+        // check no pending proposal of same type
+        $query = $system->db->query("SELECT * FROM `proposals` WHERE `end_time` IS NULL AND `village_id` = {$player->village->village_id} AND (`type` = 'offer_trade' OR `type` = 'accept_trade') LIMIT 1");
+        if ($system->db->last_num_rows > 0) {
+            return "There is already a pending trade request.";
+        }
+        // check has resources / regions
+        $is_valid_trade = self::checkTradeValid($system, $player->village->village_id, $target_village_id, $offered_resources, $offered_regions, $requested_resources, $requested_regions);
+        if (!$is_valid_trade) {
+            return "One or either village does not have the necessary items to complete the trade.";
+        }
+        // insert into DB
+        $trade_data = [
+            "offered_resources" => $offered_resources,
+            "offered_regions" => $offered_regions,
+            "requested_resources" => $requested_resources,
+            "requested_regions" => $requested_regions
+        ];
+        $trade_data = json_encode($trade_data);
+        $time = time();
+        $name = "Offer Trade: " . VillageManager::VILLAGE_NAMES[$target_village_id];
+        $type = self::PROPOSAL_TYPE_OFFER_TRADE;
+        $system->db->query("INSERT INTO `proposals` (`village_id`, `user_id`, `start_time`, `type`, `name`, `target_village_id`, `trade_data`) VALUES ({$player->village->village_id}, {$player->user_id}, {$time}, '{$type}', '{$name}', {$target_village_id}, '{$trade_data}')");
+
+        // create notification
+        self::createProposalNotification($system, $player->village->village_id, NotificationManager::NOTIFICATION_PROPOSAL_CREATED, $name);
+
+        return "Proposal created!";
+    }
+
+    private static function checkTradeValid(System $system, int $village1_id, int $village2_id, array $offered_resources, array $offered_regions, array $requested_resources, array $requested_regions): bool {
+        // get regions
+        $village1_regions = $system->db->query("SELECT `region_id` FROM `regions` WHERE `village` = {$village1_id}");
+        $village1_regions = $system->db->fetch_all($village1_regions);
+        $village2_regions = $system->db->query("SELECT `region_id` FROM `regions` WHERE `village` = {$village2_id}");
+        $village2_regions = $system->db->fetch_all($village2_regions);
+        // get resources
+        $village1_resources = self::getResources($system, $village1_id);
+        $village2_resources = self::getResources($system, $village2_id);
+        // check valid
+        foreach ($offered_resources as $resource) {
+            $resource_id = $resource['resource_id'];
+            $required_count = $resource['count'];
+            if (!isset($village1_resources[$resource_id]) || $village1_resources[$resource_id] < $required_count) {
```

Comment:
Kind of worried that someone can trade away all their resources, this seems like one of the biggest potentials for abuse. We do have other safeguards around getting Kage powers, but maybe it's best to put a cap on # of resources you can do in one trade? e.g. 10-20% of village's total resources?

Reply:
Wondering if it should be a flat value versus %, but otherwise agree on the restriction.